src/chainstate/stacks/db/accounts.rs

```diff
@@ -287,20 +279,58 @@ impl StacksChainState {
         assert!(lock_amount > 0);
 
         let mut snapshot = db.get_stx_balance_snapshot(principal);
+
+        if snapshot.has_locked_tokens() {
+            return Err(Error::PoxAlreadyLocked);
+        }
+        if !snapshot.can_transfer(lock_amount) {
+            return Err(Error::PoxInsufficientBalance);
+        }
+        snapshot.lock_tokens_v2(lock_amount, unlock_burn_height);
+
+        debug!(
+            "PoX v2 lock applied";
+            "pox_locked_ustx" => snapshot.balance().amount_locked(),
+            "available_ustx" => snapshot.balance().amount_unlocked(),
+            "unlock_burn_height" => unlock_burn_height,
+            "account" => %principal,
+        );
+
+        snapshot.save();
+        Ok(())
+    }
+
+    /// Lock up STX for PoX for a time.  Does NOT touch the account nonce.
+    pub fn pox_lock_v1(
+        db: &mut ClarityDatabase,
+        principal: &PrincipalData,
+        lock_amount: u128,
+        unlock_burn_height: u64,
+    ) -> Result<(), Error> {
+        assert!(unlock_burn_height > 0);
+        assert!(lock_amount > 0);
+
+        let mut snapshot = db.get_stx_balance_snapshot(principal);
+
+        if snapshot.balance().was_locked_by_v2() {
+            debug!("PoX Lock attempted on an account locked by v2");
+            return Err(Error::DefunctPoxContract);
```

Comment:
Would it make sense to call this error `DefunctContract`?  I suspect we'll replace other contracts this way as well as time goes on.

Reply:
I think it's more descriptive as DefunctPoxContract for now. If we use similar things for other contracts, at some point, we can change this to like `DefunctContract(QualifiedContractIdentifier)`.